Adding new RAG Chatbot example with llamaGuard and Alignscore integration

abodhankar opened the pull request:
LlamaGuard Alignscore Integration with RAG

abodhankar commented:
@drazvan for review

drazvan reviewed (changes requested):
Thanks, @abodhankar!

Generally, looks good. However, there are a few issues to solve before we'd be able to merge:
- currently there is no documentation
- the code is not a bot, so it should not be placed under `examples/bots`
- given there's also an ingestion part, this would be a better fit for a notebook; the notebook would explain the goal, the ingestion part, and then adding the guardrails; you can place it in `examples/notebooks`; 
- the config itself (without the python files) can be placed under `examples/configs`; the notebook can copy it from there; 
- including a 200 pages PDF in the repository is not recommended; can this be downloaded dynamically from somewhere? You can add this part to the notebook as well. 
- only one commit seems to be signed. Please configure signing the commits as per the contributing guidelines, and then rebase on top of the `develop` branch. 
- make sure you link to the documentation about setting up the AlignScore server. 

I'll review again once the above points have been addressed.

abodhankar commented:
@drazvan Made all the changes as per request.

Pouyanpi reviewed (commented):
Please note that, langchain_nvidia_ai_endpoints` most recent version only supports langchain 3 which is not currently supported by NeMo Guardrails. We will have it merged to the develop branch latest this week.

abodhankar commented:
@Pouyanpi the example was built using the following dependencies - 
langchain-nvidia-ai-endpoints==0.2.1
langchain==0.2.14
I can add a requirements.txt file. Do you want me to add the same in the 'pip' command in the notebook as well? But if EOW is adding the latest version compatibility. It should be fine right?

Pouyanpi commented:
@abodhankar, having the `requirements.txt` is a good idea. If it also works with `langchain 3`, we can remove it.

Please run `pre-commit` on your files by following the instructions in the [CONTRIBUTING.md](https://github.com/NVIDIA/NeMo-Guardrails/blob/eef798d86e9ab0f2f8c6d4217d23f568947b660f/CONTRIBUTING.md?plain=1#L69).

You can then run:

```sh
pre-commit run --all-files
```

The pipeline is failing because of this.

Finally, make sure to sign all of your commits. If you have GPG configured, the easiest way would be to run the following after checking out this branch:

```sh
git rebase --exec 'git commit --amend --no-edit -n -S' -i origin/main
```

Also make sure that the it works with the latest state of develop branch (it should). I will merge when the pipeline pass and the commits are signed. Thank you!

Pouyanpi commented:
Hi @abodhankar, I can fix these minor issues (running pre-commit and signing the commits) so that we can merge the PR, would it be OK for you?

drazvan commented on examples/configs/llamaguard_alignscore/config.yml:
```diff
@@ -0,0 +1,42 @@
+      The bot is knowledgeable about the company policies.
+      If the bot does not know the answer to a question, it truthfully says it does not know.
+
+sample_conversation: |
+  user action: user said "Hi there. Can you help me with some questions I have about NVIDIA AI Enterprise?"
+  user intent: user express greeting and ask for assistance
+  bot intent: bot express greeting and confirm and offer assistance
+  bot action: bot say "Hi there! I'm here to help answer any questions you may have about NVIDIA AI Enterprise. What would you like to know?"
+  user action: user said "What is offered by NVIDIA AI Enterprise?"
+  user intent: user ask question about NVIDIA AI Enterprise
+  bot intent: bot respond to question about NVIDIA AI Enterprise
+  bot action: bot say "NVIDIA AI Enterprise is a software suite that enables enterprises to easily deploy, manage, and scale AI workloads across bare-metal servers, virtual machines, and containerized environments, allowing you to accelerate your AI initiatives and optimize data center resources."
```
This sample conversation format is for Colang 2.0. However, I see that this config does not specify the Colang version, which means it will implicitly use 1.0.

abodhankar replied:
@drazvan I can mention in the Readme.md file that this is explicitly using colang1.0. Will that be good?

abodhankar replied:
@drazvan Added what I suggested in the above comment

drazvan replied:
@abodhankar : the point is that it doesn't make sense to include Colang 2.0 format sample conversation in a Colang 1.0 config.

drazvan commented on examples/configs/llamaguard_alignscore/config.yml:
```diff
@@ -0,0 +1,42 @@
+          model_name: "meta-llama/LlamaGuard-7b"
+
+rails:
+    input:
```
Reformat this YAML file to use consistent indentation.

drazvan commented on examples/notebooks/LlamaGuard_AlignScore_NIM_NeMoGuardrails.ipynb:
```diff
@@ -0,0 +1,1171 @@
+        "\n",
+        "The following is the architecture diagram for the NIM-powered RAG pipeline that we are going to build in this tutorial using NVIDIA NIM and NeMO Guardrails.\n",
+        "\n",
+        "![Arch Diagram](https://drive.google.com/file/d/17iUW3ReMq5h5aFAb5DvZbtN8B_iEcUVf/view?usp=drive_link)\n"
```
This diagram is not publicly accessible. Please include the diagram in the repo, next to the notebook, if needed.

abodhankar commented:
@drazvan @Pouyanpi Finished doing all the changes.

drazvan reviewed (changes requested):
- For the architecture diagram, update the link to point to the image; it still points to the GDrive one;
- `wget` may not work when running the on mac (as it was the case for me); replace with `curl` command 
	- `!curl -o ./kb/data.pdf https://docs.nvidia.com/ai-enterprise/latest/pdf/nvidia-ai-enterprise-user-guide.pdf`
- The PDF https://docs.nvidia.com/ai-enterprise/latest/pdf/nvidia-ai-enterprise-user-guide.pdf is no longer accessible
- remove `.ipynb_checkpoints` folder;
- lowercase for notebook filename for consistency with the other notebooks;
- `pypdf` package needs to be installed; following the tutorial as is results in import error.
- the formatting in the "To understand the config ...." is off (I think the parameters should be a bulleted list)
- there are inconsistencies in the formatting of YAML snippets (the `%%writefile config/config.yml` sections)

abodhankar commented:
I have created a directory for the notebook that has the following - 
1. Arch Diagram
2. kb
3. final notebook

Now Let me address the comments here - 
1. For the architecture diagram, update the link to point to the image; but I still couldn't render it well it still points to the GDrive one - I wasn't able to render it properly in this in the notebook. So, added the image with the notebook.
2. **wget may not work when running the on mac (as it was the case for me); replace with curl command
!curl -o ./kb/data.pdf https://docs.nvidia.com/ai-enterprise/latest/pdf/nvidia-ai-enterprise-user-guide.pdf**
---- For this I checked that the pdf is moved to a different location. So, for not having future issues, I have added the kb instead. 
3. **The PDF https://docs.nvidia.com/ai-enterprise/latest/pdf/nvidia-ai-enterprise-user-guide.pdf is no longer accessible**
---- Check point 2 above
4.  Removed .ipynb_checkpoints folder - DONE
5. lowercase for notebook filename for consistency with the other notebooks; DONE
6. **pypdf package needs to be installed; following the tutorial as is results in import error** the error was because the pdf was not in the location specified before, so it re-directed to a webpage and downloaded content into pdf causing rendering issues and hence pypdf issue - SORTED NOW
7. the formatting in the "To understand the config ...." is off (I think the parameters should be a bulleted list) - DONE
8. there are inconsistencies in the formatting of YAML snippets (the %%writefile config/config.yml sections) - ADDRESSED

Pouyanpi commented:
@abodhankar Should this PR be closed or is it still pending?
